tests/test_subtractTask.py

```diff
@@ -0,0 +1,459 @@
+        # There should be no NaNs in the image if we convolve the template with a buffer
+        self.assertTrue(np.all(np.isfinite(output.difference.image.array)))
+        # Mean of difference image should be close to zero.
+        nGoodPix = np.sum(np.isfinite(output.difference.image.array))
+        mean_error = np.nanstd(science.image.array)/np.sqrt(nGoodPix)
+        self.assertFloatsAlmostEqual(np.mean(output.difference.image.array), 0, atol=5*mean_error)
+        # stddev of difference image should be close to expected value.
+        self.assertFloatsAlmostEqual(np.std(output.difference.image.array), np.sqrt(2)*noiseLevel, rtol=0.1)
+
+    def test_symmetry(self):
+        """Test that convolving the science and convolving the template are
+        symmetric: if the psfs are switched between them, the difference image
+        should be nearly the same.
+        """
+        noiseLevel = 1.
+        # Don't include a border for the template, in order to make the results
+        #  comparable when we swap which image is treated as the "science" image.
+        science, sources = makeTestImage(psfSize=2.0, noiseLevel=noiseLevel,
+                                         noiseSeed=6, templateBorderSize=0)
+        template, _ = makeTestImage(psfSize=3.0, noiseLevel=noiseLevel,
+                                    noiseSeed=7, templateBorderSize=0)
+        config = subtractImages.AlardLuptonSubtractTask.ConfigClass()
+        config.mode = 'auto'
+        config.doSubtractBackground = False
+        task = subtractImages.AlardLuptonSubtractTask(config=config)
+
+        science_better = task.run(template, science, sources)
+        template_better = task.run(science, template, sources)
+
+        delta = template_better.difference.clone()
+        delta.image -= science_better.difference.image
+        delta.variance -= science_better.difference.variance
+        delta.mask.array -= science_better.difference.mask.array
+
+        # Mean of delta should be very close to zero.
+        nGoodPix = np.sum(np.isfinite(delta.image.array))
+        mean_error = np.nanstd(science.image.array)/np.sqrt(nGoodPix)
+        self.assertFloatsAlmostEqual(np.nanmean(delta.image.array), 0, atol=5*mean_error)
+        # stddev of difference image should be close to expected value
+        self.assertFloatsAlmostEqual(np.nanstd(delta.image.array), 2*np.sqrt(2)*noiseLevel, rtol=.1)
+
+    def test_few_sources(self):
+        """Test with only 1 source, to check that we get a useful error.
+        """
+        xSize = 256
+        ySize = 256
+        science, sources = makeTestImage(psfSize=2.4, nSrc=1, xSize=xSize, ySize=ySize)
+        template, _ = makeTestImage(psfSize=2.0, nSrc=1, xSize=xSize, ySize=ySize)
+        config = subtractImages.AlardLuptonSubtractTask.ConfigClass()
+        task = subtractImages.AlardLuptonSubtractTask(config=config)
+        with self.assertRaises(lsst.pex.exceptions.Exception):
```

Comment:
We don't want to check for a generic exception here, we want to check for a specific one, and that the error message is useful (using `self.assertRaisesRegex`). I don't know whether the existing behavior is actually what we want, or whether the subtractTask should do a `raise SomeException("not enough sources") from e` or something like that: I recall that the exception we get in this case is not very helpful.

Reply:
I made the test more precise as you suggested, but did not fix up the underlying error message. If we have a chance, it would be a good idea to try to clean up the code that solves for the matching kernel, including better error messages.